ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1881,18 +2057,318 @@ void shutdown() throws InterruptedException {
+        if (replicaToThrottledTill.containsKey(remoteReplicaInfo)
+            && replicaToThrottledTill.get(remoteReplicaInfo) > time.milliseconds()) {
+          nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+          throttleCount.inc();
+        }
+      });
+
+      // remove all replicas of corresponding group id if group is in iteration
+      if (inflightGroups.containsKey(groupId)) {
+        logger.trace("Thread name: {} group {} is in iteration", threadName, groupId);
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightGroups.get(groupId).remoteReplicaInfos);
+      }
+
+      // remove all replicas of corresponding standby group if group is in iteration
+      if (inflightGroups.containsKey(standByGroupId)) {
+        logger.trace("Thread name: {} stand by group {} is in iteration", threadName,
+            standByGroupId);
+        nonIteratingNonPendingRemoteReplicas.removeAll(
+            inflightGroups.get(standByGroupId).remoteReplicaInfos);
+      }
+
+      // filter active and stand by replicas
+      filterRemoteReplicasToReplicate(new ArrayList<>(nonIteratingNonPendingRemoteReplicas), activeReplicas,
+          standbyReplicasWithNoProgress);
+    }
+
+
+    /**
+     * Iterates over {@link #groupIdToRemoteReplicaMap} all potential groups,
+     * tries to create active groups and adds new groups to {@link #inflightGroups}
+     * also adds replicas which have timed out on stand by
+     * to {@link #standByReplicaQueue} for being picked by standby groups
+     */
+    private void createActiveGroups(){
+      logger.trace("Thread name: {} Trying to create new active groups", threadName);
+      for (Map.Entry<Integer, List<RemoteReplicaInfo>> entry : groupIdToRemoteReplicaMap.entrySet()) {
+        int groupId = entry.getKey();
+        logger.trace("Thread name: {} Processing for group id {}", threadName, groupId);
+        List<RemoteReplicaInfo> remoteReplicas = entry.getValue();
+        DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+        // get stand by group id for the data node
+        int standByGroupId = standByGroupIds.get(remoteNode);
+
+        List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+        // generate active and standby replicas with no progress
+        getActiveAndStandByNoProgressReplicas(activeReplicas, standbyReplicasWithNoProgress, remoteReplicas,
+            standByGroupId, groupId);
+        logger.trace("Thread name: {} Active replicas are {}, standby Replicas with no progress are {}", threadName,
+            activeReplicas, standbyReplicasWithNoProgress);
+
+        // if the current group is not in iteration, and there are active replicas, we will start a new group
+        // and increase iteration count if we are able to create
+        if (!inflightGroups.containsKey(groupId)) {
```

Comment:
We should check if the group id is already in the inflight map before getting the active replica and standby with no progress. This method would be called quite often, every time in the do-while loop, we would try to poll and create new groups, which means most of the time, the group is still inflight. If we check the inflight group first, we would avoid the logic above most of the time.

Reply:
For this group id, all replicas may not be in the inflight group, some may have been stand by and timeout can happen any time also.

Reply:
For the active replicas that are not in the group, they will not be added to the group anyway, since the group is still inflight. Even we find more active replicas this time, they are still going on next iteration. 
For timedout standby replicas, it's the same case. Even we find more timed out standby replicas, if the standby group is still inflight, they will not be added to the current group anyway. 

That's why i suggest separate the function to add active replicas and timed out standby replicas. They belong to different groups and will not be reevaluated at the same time.

Reply:
If we do find timedout standby replicas, we are adding to queue in  code below. and whenever standBygroup is free they will be picked up.
Sure, we can  add to queue when active group is done, but here i am checking if any replica not added to active group has timedout, then i can add to queue as soon as possible.
We will add standby replica to queue sooner in current approach